webapp/lib/controller.dart

```diff
@@ -267,48 +277,86 @@ bool isProjectStale(model.NeedsReplyData projectData) {
   }
 }
 
-void setupProjectTimer(model.NeedsReplyData projectData, [bool stale = false]) {
-  projectTimers[selectedProject]?.cancel();
+String getWatchdogTimer(Object data) {
+  if (data is model.NeedsReplyData) {
+    return  selectedProject;
+  } else if (data is model.SystemMetricsData) {
+    return SYSTEM_METRICS_ROOT_COLLECTION_KEY;
+  } else {
+    throw new model.ModelDoesNotExistException('Model does not exist');
+  }
+}
+
+void setupWatchdogTimer<T>(T latestData, [bool stale = false]) {
+  var data;
+  if (latestData is model.NeedsReplyData) {
+    data = latestData as model.NeedsReplyData;
+  } else if (latestData is model.SystemMetricsData) {
+    data = latestData as model.SystemMetricsData;
+  }
+
+  watchdogTimers[getWatchdogTimer(data)]?.cancel();
 
   if (stale) {
-    projectTimers[selectedProject] = null;
+    watchdogTimers[getWatchdogTimer(data)] = null;
   } else {
-    var timeToExecute =  projectData.datetime.add(Duration(hours: 2));
+    var timeToExecute =  data.datetime.add(Duration(hours: 2));
     var now = new DateTime.now();
     var duration = timeToExecute.difference(now);
     var timer = new Timer(duration, () {
-      view.contentView.stale = true;
+      if (data?.project == selectedProject) {
```

Comment:
You access properties of `data` in a couple of places in this function. Since `data` (and `latestData` before it) shouldn't really be null for this function to work, I think it might be easiest to assume in this function that they're non-null variables and don't do any checks, but check instead where it get computed (ie. in `checkNeedsReplyMetricsStale` and `checkSystemMetricsStale`).

If we get a null data point in the middle of the application run, then we have a problem and we should handle it in some way rather than absorb it, which is what we seem to be doing here?

Reply:
Now doing the data empty check before all this as we discussed.